pkg/querier/cardinality/request.go

```diff
@@ -0,0 +1,225 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package cardinality
+
+import (
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql/parser"
+	"golang.org/x/exp/slices"
+)
+
+const (
+	RequestTypeLabelNames  = RequestType(iota)
+	RequestTypeLabelValues = RequestType(iota)
+
+	minLimit     = 0
+	maxLimit     = 500
+	defaultLimit = 20
+
+	stringSeparator = '|'
+)
+
+type RequestType int
+
+type LabelNamesRequest struct {
+	Matchers []*labels.Matcher
+	Limit    int
+}
+
+// Strings returns a full representation of the request. This returned string can be
+// used to uniquely identify the request.
+func (r *LabelNamesRequest) String() string {
+	b := strings.Builder{}
+
+	// Add matchers.
+	for idx, matcher := range r.Matchers {
+		if idx > 0 {
+			b.WriteRune(',')
+		}
+		b.WriteString(matcher.String())
+	}
+
+	// Add limit.
+	b.WriteRune(stringSeparator)
+	b.WriteString(strconv.Itoa(r.Limit))
+
+	return b.String()
+}
+
+func (r *LabelNamesRequest) RequestType() RequestType {
+	return RequestTypeLabelNames
+}
+
+// TODO doc
+// TODO test (GET and POST)
+func DecodeLabelNamesRequest(r *http.Request) (*LabelNamesRequest, error) {
+	var (
+		parsed = &LabelNamesRequest{}
+		err    error
+	)
+
+	err = r.ParseForm()
+	if err != nil {
+		return nil, err
+	}
+
+	parsed.Matchers, err = extractSelector(r)
+	if err != nil {
+		return nil, err
+	}
+
+	parsed.Limit, err = extractLimit(r)
+	if err != nil {
+		return nil, err
+	}
+
+	return parsed, nil
+}
+
+type LabelValuesRequest struct {
+	LabelNames []model.LabelName
+	Matchers   []*labels.Matcher
+	Limit      int
+}
+
+// Strings returns a full representation of the request. This returned string can be
+// used to uniquely identify the request.
+func (r *LabelValuesRequest) String() string {
+	b := strings.Builder{}
+
+	// Add label names.
+	for idx, name := range r.LabelNames {
+		if idx > 0 {
+			b.WriteRune(',')
```

Comment:
When new characters are allowed in label names, it may be possible to construct conflicting requests (eg. request with label names `["a", "b"]` and request with single label `["a,b"]` would have same representation).

It may be safer to use `\0` as delimiter or some other value that will "for sure" never appear in label names or matchers.

Reply:
Right. I've used `\0` and `\1` as param and values separator.